### ISSUE TITLE
Regarding range_plus_one

### ISSUE THREAD

**Issue by leonardo-m** [rust-lang/rust-clippy#8317]
### Summary

This code:

```rust
#![warn(clippy::pedantic)]

fn foo(i: u32) {
    for i in 0 .. i + 1 {
        println!("{}", i);
    }
}
```

Gives:

```rust
warning: an inclusive range would be more readable
 --> src/main.rs:4:14
  |
4 |     for i in 0 .. i + 1 {
  |              ^^^^^^^^^^ help: use: `0..=i`
  |
note: the lint level is defined here
 --> src/main.rs:1:9
  |
1 | #![warn(clippy::pedantic)]
  |         ^^^^^^^^^^^^^^^^
  = note: `#[warn(clippy::range_plus_one)]` implied by `#[warn(clippy::pedantic)]`
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#range_plus_one
```

But ..= isn't performant, especially in inner loops. So for me this seems a bad warning.

### Reproducer

I tried this code:

```rust
<code>
```

I saw this happen:

```
<output>
```

I expected to see this happen:

**Comment by blyxyas** [comment]
[Godbolt](https://rust.godbolt.org/#z:OYLghAFBqd5TKALEBjA9gEwKYFFMCWALugE4A0BIEAZgQDbYB2AhgLbYgDkAjF%2BTXRMiAZVQtGIHgBYBQogFUAztgAKAD24AGfgCsp5eiyahSAVyVFyKxqiIEh1ZpgDC6embZMQAJnLOAGQImbAA5TwAjbFIQADZyAAd0JWIHJjcPL19E5NShIJDwtiiY%2BJtsOzSRIhZSIgzPbx5rbFt7IWraogKwyOiDSy6GrObBup6ikqkASmt0M1JUTi4AUh8AZhWAVgAhJnQAfTZjYEZtgBEVrQBBBLMIgGoaJif0dAgqB7N1n2mHlYA7DsrtcHmDXqQHi9gg8tAA6OEEf4%2BHY8f5AkEJUjBIj0JhgSBrHyAvaAy4%2BX4rdbAm7g9GXG5krizejcLb8by8fjobgAJQsRAeSnmi2wyPWfHIRG0zNmAGsQFstIZuNJ%2BGxFcrOTpyDyuPwlCBldKuDpZnBYCgMDh8MQyJRqHRGKwONxJYJhGIJJwZHJhMo1JpTXpskYTCBzJYWm00k4mK53I1soFgr1iv14kkUu10omsn4s3kmBM%2BqVoxUc506sNvH5ypUOjVxqnJhnrE36nna%2B2uiX06VZsKFktuGtNrt9kcTmctgzbvcni9BO9Pt9fuiaaDwYJIdCXvC4VTzkiSZjscI8QSIESSUwyUTplTN3T74yAedmSquOzyNruXyBSFEUlnFHx%2BBNM1ZiQbAWBwGIIBZVV1U1X8ZV1bgDSNKUZWmeUUNZLh1g5NC9XAnDyAtZA0HQNgEgYaIHQgDBaPomJSB4AAOHhlSdIhokNCAIjQiJglqABPN1%2BBE1hSDEgB5CJdArSTyGYjhhDkph6Ak4NyBwCIzGAFwJHoQ0uT07BjhMSRdMIUgKwIAA3bAzJ1bB1AqMw%2BJUnFWjQ%2BgCAiUhxLcHA0KIbENXM5zSAiZJsHOSzp2CUBg1mGgjGAJQADUCGwAB3OSEmYFSPVEcRJF9MqAw0ND9GaMNTAFQxAsNSBZnQBIczMgBaNwup6xhnPoI8iN1GLsRwNqoBdThsEIHNyEciQzGWHwtB8PhcPLBtvAgZwawMFNClLAxCxzQ7mnOtI%2BymUZWgcxshi7AYHt2qtuhbU7Rg7S6e2bE7%2BxmOZhx9RDv2I3S9QeSMiFQB4eDhLi4S0B4IFtEhITHHhpjItK8J8LYvzVVCoYw6wsIgz9KIgK0aLoxgKCoJj6dYiNOO4gQGD40gBKE3TpPElTBdkhSlLsFS1OYIhNO0tD9MM4z6FMlScCs05lh1OyHOc1z%2BHczzvPM3yCJ1AKgpCrBNfAyKVJiuKVES9WAvDNKBEynK8sK4rOXdeQvUq2RqpUWrdP0PxGojZrzemjqurSXr%2BqIQbsGG0buQmggpvgCBZpAeaiyWla1o2rbBzenM4wTTJvA4/x41u/ofEzXILpeuvrvyL6gebnbK1%2B9u%2B6qDtG5iXuxk7GuQDrifR98WJB2Azhfi/H8/3QrgYYFeHEeR1H0YWshQNx7D8fIaDYP6BDyAVQnichnVSIp40cPBsDScf8mqe2mKUkcaQgA%3D)

I'm not sure why ..= produces different output than ..n + 1, we'll have to fill a report on the compiler directly, as this isn't a Clippy issues.

**Comment by saethlin** [comment]
This lint was downgraded to pedantic in response to https://github.com/rust-lang/rust-clippy/issues/2217. So I suppose this issue is about whether it's okay for clippy to suggest deoptimizations in pedantic.
